cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetInstanceTable.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2014-2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.service.mds;
+
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.StructuredTableContext;
+import co.cask.cdap.spi.data.TableNotFoundException;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Fields;
+import co.cask.cdap.spi.data.table.field.Range;
+import co.cask.cdap.store.StoreDefinition;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset instances metadata store
+ */
+public final class DatasetInstanceTable {
+  private static final Gson GSON = new Gson();
+
+  private StructuredTable table;
+
+  public DatasetInstanceTable(StructuredTableContext context) throws TableNotFoundException {
+    table = context.getTable(StoreDefinition.DatasetInstanceStore.DATASET_INSTANCES);
+  }
+
+  @Nullable
+  public DatasetSpecification get(DatasetId id) throws IOException {
+    Optional<StructuredRow> row =
+      table.read(
+        ImmutableList.of(Fields.stringField(StoreDefinition.DatasetInstanceStore.NAMESPACE_FIELD,
+                                            id.getParent().getEntityName()),
```

Comment:
you can call id.getNamespace() and id.getDataset(). I think that is clearer to read.